Block Editor Settings: use v2 DropdownMenu

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

⚠️ ⚠️ ⚠️  WIP ⚠️ ⚠️ ⚠️ 

Part of #50459

## What?
<!-- In a few words, what is the PR actually doing? -->

This PR refactors the block toolbar's "settings" dropdown menu — the menu that is toggled by pressing the last button (labelled "Options" and associated to a vertical "..." icon.

The refactor involved migrating from the legacy `DropdownMenu` component to the new version (currently still experimental and only available as a locked private API of the `@wordpress/components` package).

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The main advantage of migrating to the new version of the component is the support for nested menus.

This PR also represents the first "real world" usage of the experimental component (ie. not isolated in Storybook), which will expose a number of issues related to integrating the component with the existing set of components and constraints of the block editor.

**Is this PR including breaking changes?**

No, this PR does not include breaking changes. The dropdown menu and all its items have been refactored to use the new experimental version of the `DropdownMenu` component, without otherwise changing any public-facing APIs (including the `fillProps` passed to the different slotfills).

Even if a third-party developer was to add a legacy `MenuItem` to the block settings dropdown via the `BlockSettingsMenuControls` slotfill, the `MenuItem` would continue to render without any errors, and clicking on the item would still cause the expected results. The only difference may be visual, as the new dropdownmenu components have slightly different styles and spacing.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

These are the changes being proposed in this PR. As the PR grows in scope, we may decide to split some of the proposed changes to separate, smaller PRs.

### 📝 Changes

<details>
<summary>Expand to show the detailed list of changes</summary>

**General changes:**
- Swapped legacy components (namely `DropdownMenu`, `MenuGroup`, `MenuItem`) with the new equivalent experimental components. Main differences:
  - Instead of `toggleProps` prop, the component requires a trigger to be explicitly passed via the `trigger` prop. I've therefore passed a `Button` component as the trigger and forwarded the `trigger` props to it
  - Instead of the `menuProps` prop, the component can accept certain props directly (like `onKeyDown`)
  - The component doesn't accept the `disableOpenOnArrowDown` prop. I've implemented the same functionality by controlling the `DropdownMenu` component with internal state and added a keyboard event listener on the trigger to `preventDefault` when the arrow down key is pressed.
  - The new `DropdownMenuItem` component doesn't accept a `shortcut` prop. Instead, I've created a custom `Shortcut` component and passed it via the `suffix` prop
  -  On a similar note, the new `DropdownMenuItem` component doesn't accept an `icon` prop. Instead, I've used the `prefix` prop and passed the icon via the `Icon` component
  - Renamed `onClick` props to `onSelect` for the new `DropdownMenuItem`.
  - The old `DropdownMenu` component accepted, as `children`, a render function. This function accepted an object containing the `onClose` callback, which could be used to programmatically close the dropdown menu. The new `DropdownMenu` component doesn't expose this functionality, but an equivalent callback has been created and passed to the children components for backward-compat reasons.
  - The new `DropdownMenuItem` component automatically closes the dropdown menu when clicked. This meant that a lot of the calls to the `onClose` callback have been removed, and that explicit `event.preventDefault()` calls have been added where the menu needs to stay open after clicking the menu item.
  - The new `DropdownMenuGroup` component doesn't render a line to visually demarcate the group. Instead, I used the new `DropdownMenuSeparator` component.
- Refactored the "copy" functionality away from the `useCopyToClipboard` hook, and instead implemented a custom solution using native web APIs instead (more details on this below)
- Overhauled some classnames and styles associated to the dropdown. In particular, styles the `packages/block-editor/src/components/block-settings-menu/style.scss` were deleted, since they were targeting the legacy dropdown. As a consequence, I also removed all occurrences of the `block-editor-block-settings-menu` and `block-editor-block-settings-menu__popover` classes from the repository.
  - Instead, I've added new classnames for the button trigger, which seem to be needed in order to hide the dropdown in react-native environments. While doing so, I've created three different classnames, instead of reusing the same classname across packages.
  - Another consequence of this change is that one of the puppeteer e2e test required refactoring.
- Added styles for the dropdown menu's content to allow for the menu to resize and show a scrollbar when there isn't enough space for it to display fully
- The `useCopyToClipboard` hook has been updated to support an `container` argument, useful to circumvent the focus trap added by the `DropdownMenu` component

**Changes applied to the experimental `DropdownMenu` component:**
- Added support for the `onKeyDown` prop on the root `DropdownMenu` component. This is to support the use-case of the block toolbar settings menu, which uses the prop to support keyboard shortcuts ([source](https://github.com/WordPress/gutenberg/pull/50422#discussion_r1228012486))
- Changed cursor styles to show a pointer over the menu items (apart from disabled items)
- Added the same `z-index` for content wrapper elements as the legacy `Popover` menu, to avoid stacking order bugs when other modals are opened.
- Added support for a few pointer-related callback props on `DropdownMenuItem`. This is to support the use-case of the "select parent block" menu item, which is supposed to highlight the parent block when hovering the menu item
- Added support for the `href`, `rel`, and `target` props on `DropdownMenuItem`. This is to support the use-case of the "Manage reusable blocks" menu item, which is supposed to link away from the editor. When the `href` prop is not empty, the `DropdownMenuItem` automatically renders as an `<a />` tag instead of a `<div />`. Some extra styles were also necessary to override the default Gutenberg styles associated with anchor tags.
- Allow passing a few more HTML attributes to `DropdownMenu`, like `id`, `aria-labelledby`, `className`. These attributes are often used by consumers of the component, and were necessary in ensuring that the trigger's id was passed correctly to the menu's context in this specific PR instance
- Added some additional styles to the content wrappers to make legacy `MenuItem` components look better

**Changes applied to the `Modal` component:**
- Added `pointer-events: all` to the modal wrapper. This is to override styles added by the Radix DropdownMenu, which sets `pointer-events: none` when opened in a modal way. Without the style override, the modal would not receive pointer events.
  - Note that this doesn't seem to be enough in order to make modals opened _on top_ of the `DropdownMenu` behave as expected.

</details>

### 🛑 Pending issues

<details>
<summary>The new DropdownMenu component interferes with Modal components opened on top of the dropdown</summary>

Due to how the new `DropdownMenu` component is implemented, Modals rendered on top of open dropdown menus don't react to pointer events and don't get keyboard focus.

I was able to (potentially) solve the pointer events problem (by forcing `pointer-events: all` on the modal), although that solution would require more testing to make sure it's robust (for example, what happens if a modal DropdownMenu is opened inside a modal?).

What I haven't been able to solve so far regards keyboard focus. The new `DropdownMenu` component, in fact, seems to trap keyboard focus on the dropdown. And therefore, when opening a modal on top of an opened dropdown menu, the keyboard focus stays on the dropdown menu (this can be verified because, when typing letters, the dropdown menu still highlights the matching menu item.

It looks like we'd need to use the `Dialog` component from Radix to have modals work as expected: https://codesandbox.io/s/dropdownmenu-dialog-items-forked-9sy6j8?file=/src/App.js
</details>

### 💬 Pending discussions

<details>
<summary>Where and how to implement the resize behavior when meeting the edge of the viewport</summary>

Radix's `DropdownMenu` offers this behavior by exposing the `--radix-dropdown-menu-content-available-height` CSS variable, and letting the consumer of the component actually implement the resizing behaviour. This is what I've done so far in this PR, but we should ask ourselves a few questions:

- Should we offer this same functionality directly in the `DropdownMenu` component? If so, we should think carefully about how to expose it and future-proof it against potential radix changes
- In case we keep it out of `DropdownMenu`, would it be ok that consumers of the component use a CSS variable that clearly states '"radix" ? It would be good to keep radix as an implementation detail.
</details>

<details>
<summary>Decide the best way to allow `DropdownMenuItem` to render as an anchor tag</summary>

Currently, the `DropdownMenuItem` exposes the `href` prop and internally decided to render an `<a />` or a `<div />`. This approach is simple and mimics the APIs of the legacy `MenuItem` and the `Button` component, but it lacks in scalability (what if we needed to render something else as a menu item?)

Alternative approaches:
- expose the `asChild` prop, allowing consumers of the component to pass whatever component they want
- always leave `asChild` set to `true` in the internal implementation, requiring consumers of the component to always specify a "menu item wrapper" element
</details>

<details>
<summary>Reflect on which props we should expose on `DropdownMenu` and `DropdownMenuItem`</summary>

Using the components in a real-world scenario highlighted the need for consumers of the components to pass certain HTML attributes, like `className`, `onKeyDown`, and `aria-labelledby`.

So far, our strategy on these components has been to reduce the amount of props exposed as much as possible, but I wonder if we should reconsider that and allow a wider range of props (ie. standard HTML attributes?)
</details>

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Make sure that:

- the block toolbar settings dropdown menu opens/closes correctly
- the block toolbar settings dropdown menu positions itself correctly when meeting the edge of the viewport (including nested menus?)
- modal behavior: when the new dropdown is visible, the rest of the page is not interactive (ie. does not scroll / react to pointer events / content is not accessible)
- all menu items are shown/hidden following the correct logic
- all menu items behave as expected when clicked (both the functionality that they execute when clicked, and the fact that the dropdown menu is closed or stays open)
- make sure that all other v1 dropdown menus are still working as expected, especially the ones that share a lot of the same functionality (like the one in the list view)
- keyboard shortcuts work as expected (including [onKeyDown ev listener](https://github.com/WordPress/gutenberg/pull/50422#pullrequestreview-1476985805))
- All the menu items refactored to the new dropdownmenu component are not rendered in other legacy menus

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

| `trunk` | this PR (wip) |
|---|---|
| ![Screenshot 2023-06-12 at 12 02 40](https://github.com/WordPress/gutenberg/assets/1083581/2b823efb-da40-43b3-b918-8f693c739148) | ![Screenshot 2023-06-12 at 13 35 45](https://github.com/WordPress/gutenberg/assets/1083581/03f80f8a-1005-47e4-aed9-7276fb3f3f08) |

ciampo committed:
Start using dropdownmenu v2 in block settings dropdown

ciampo committed:
Add Shortcut component

ciampo committed:
Refactor block settings dropdown menu items to use dropdownmenu v2

ciampo committed:
Add notes about duplicate implementations/classnames

ciampo committed:
Convert more menu items to use the dropdownmenu v2 components

ciampo committed:
Convert menu items from the reusable-blocks package to use the dropdownmenu v2 components

ciampo committed:
Add separator to BlockSettingsMenuControls

ciampo committed:
Simplify Shortcut suffix component

ciampo committed:
Refactor more menu items rendering inside `BlockSettingsMenuControls`

ciampo committed:
Remove onKeyDown menu props (it doesn't seem to be needed?)

ciampo committed:
Migrate from calling onClose to calling preventDefault() when needing to keep the dropdown open

ciampo committed:
Fix broken copy/paste by using native clipboard APIs instead of external library

ciampo committed:
Remove unnecessary wrapper functions, unused imports, clean up code, onClick => onSelect

ciampo committed:
Add default z-index

ciampo committed:
Capture pointer events in modals

ciampo committed:
DropdownMenuV2: accept onKeyDown prop

ciampo committed:
Restore onKeyDown ev listener on the block settings dropdown

ciampo committed:
Add a few pointer-related props to DropdownMenuItemProps

ciampo committed:
Destructure and properly allocate more props in Block settings menu

ciampo committed:
DropdownMenuItem cursor styles

ciampo committed:
Clean up classnames:
- remove duplicate block-editor-block-settings-menu instances in other packages
- remove associated styles (which were not having any effects)
- add new classnames for the triggers (different for each dropdownmenu instance)
- use the new classnames in the react native style overrides

ciampo committed:
Temp fix for e2e test

ciampo committed:
Refactor block settings dropdown to controlled mode. This helps with:
- stopping the propagation of the ESC keydown, so that only the dropdown closes and the toolbar stays open
- restoring passing an onClose callback to slotfills, including the create reusable pattern modal

ciampo committed:
Support rendering anchor tags in DropdownMenuItem

ciampo committed:
Remove done  comment

ciampo committed:
Avoid unnecessary function

andrewserong commented:
I know this is still a WIP, but just thought I'd take it for a very quick spin. It looks like this is also going to resolve #50988 from early testing (CMD+click with another browser window no longer results in multiple block settings dropdown menus displaying), so I'm quite excited for this one! 😄

![2023-06-15 14 31 01](https://github.com/WordPress/gutenberg/assets/14988353/6fb9fad9-450d-42f1-be66-ed25c7c7b691)

ciampo committed:
Set the correct value for aria-labelledby, update e2e tests

ciampo committed:
Fix unit test by wrapping the BlockModeToggle in dropdownmenu

ciampo committed:
Sync package-lock.json

ciampo committed:
Cleanup code

ciampo committed:
Back compat styles for legacy menuitems

ciampo commented:
Update:

After working on this PR pretty much for the whole week, I've managed to iron out most wrinkles, but I've hit a bit of a wall: **modals opened on top of the dropdown** (for example, the one for creating reusable blocks, or for locking/unlocking a block) **don't work properly.** (as also highlighted by e2e test failures).

This is a consequence of the `modal` nature of the new experimental `DropdownMenu`, which causes the component to trap focus and disable pointer events when opened.
`Modal` components displayed on top of `DropdownMenu` don't receive keyboard focus, making them not interactive. I'm currently in the process of looking for solutions to this problem, but so far the only potential solution seems to be to refactor our `Modal` component to use Radix's `Dialog`, which is able to communicate with Radix's `DropdownMenu` and handle focus correctly.

A possibility that I briefly explored is to close the `DropdownMenu` component when opening a `Modal`, but that approach also has its flaws:
- we would still have focus management issues, since the `DropdownMenu` component would normally restore focus on its trigger when closed, thus stealing it from the `Modal`
- In the react tree, `Modals` are usually rendered as children of `DropdownMenu`; therefore, closing the dropdown menu also instantly hides the modals. Avoiding this situation would require a large re-architecture and would likely cause breaking changes for the third-party developers.

I've also explored setting the `DropdownMenu` as non-modal:
- it would only partially solve the compatibility issues with our `Modal` component
- it would cause a series of other issues, like:
  - the one mentioned by Andrew in [his comment above](https://github.com/WordPress/gutenberg/pull/51400#issuecomment-1592341706)
  - the dropdown menu would not sync its position correctly with the block toolbar as the page scrolls (this can happen when the dropdown menu is not modal)
  - probably more incompatibility issues with our `Popover`-based components, which we haven't discovered yet


We always knew that there was a chance for this kind of issue when introducing a new component that doesn't rely on our internal `Popover`, and this PR is definitely informing us on how to best tackle them.

ciampo committed:
Resize dropdown menu + show scrollbars when there's not enough space

ciampo committed:
Forward ref on DropdownMenu component

ciampo committed:
Allow specifying container on useCopyToClipboard hook

ciampo committed:
Revert to using useCopyToCliboard hook

diegohaz commented:
> This is a consequence of the `modal` nature of the new experimental `DropdownMenu`, which causes the component to trap focus and disable pointer events when opened.
> `Modal` components displayed on top of `DropdownMenu` don't receive keyboard focus, making them not interactive. I'm currently in the process of looking for solutions to this problem, but so far the only potential solution seems to be to refactor our `Modal` component to use Radix's `Dialog`, which is able to communicate with Radix's `DropdownMenu` and handle focus correctly.

I'd suggest disabling the default modal behavior and implementing a custom modal logic that considers nested/third-party dialogs. Even though you can change all dialogs in the app to use Radix's Dialog, you can't control how third-party dialogs will render. This includes browser extensions users may have installed, like Google Translate, Grammarly, 1Password, etc., but also other persistent widgets that should be accessible in parallel (like floating chat widgets).

Here's how we fixed this on Ariakit with a brief explanation: https://github.com/ariakit/ariakit/pull/2339

diegohaz commented:
> Even if a third-party developer was to add a legacy `MenuItem` to the block settings dropdown via the `BlockSettingsMenuControls` slotfill, the `MenuItem` would continue to render without any errors, and clicking on the item would still cause the expected results. The only difference may be visual, as the new dropdownmenu components have slightly different styles and spacing.

I'm also curious to learn how this works. As far as I know, SlotFill won't automatically provide context to elements rendered outside the original tree. If Radix uses React Context to control its menu items, the context should be manually re-created in the fill tree, otherwise things like roving tabindex may not work as expected. That's what the Block Toolbar component does.

---

Oh, I think I got it. The dropdown slots aren't using the `bubblesVirtually` prop. The block toolbar slot does. Is migrating all slots to `bubblesVirtually` still the plan? If so, things might break for third-party components that depend on an internal React context like Radix. [I wrote about it a few years ago](https://twitter.com/diegohaz/status/1229499934845607936).

I created this example using Ariakit Menu and the `Modal` component from `@wordpress/components` that works with `SlotFill` independently of the `bubblesVirtually` prop: https://stackblitz.com/edit/ariakit-nested-menu-wordpress?file=menu-nested%2Fmenu.tsx,menu-nested%2Findex.tsx

Some things to note:

- It works with the WordPress modal component and most third-party dialogs.

- [The modal behavior on Ariakit is similar to the native dialog element](https://ariakit.org/examples/dialog-radix#focus-trap-elements). Keyboard users can access the browser chrome the same way as mouse users do.

- Screen reader users on touch devices can't press <kbd>Esc</kbd> to close the menu. If you don't explicitly render a close menu button within the modal menu, those users will be trapped there. Their only options will be either reloading the page or clicking on a menu item (that does another thing in addition to closing the menu). That's why Ariakit will automatically render a visually hidden dismiss button inside the modal menu.

- Currently, Ariakit uses the store to control most of the menu behaviors, but nested menus are controlled using React Context, just like Radix. Nested menus inside `SlotFill` with `bubblesVirtually` may not work as expected (they will just be a dropdown menu inside another dropdown menu). But supporting nested menus that aren't nested in the React tree is in the plans. This should be a relatively easy implementation, and the API would be as simple as this:

    ```jsx
    const parent = useMenuStore()
    const nested = useMenuStore({ parent })
    ```

ciampo commented:
Hey @diegohaz , thank you for chiming in!

> I'd suggest disabling the default modal behavior and implementing a custom modal logic that considers nested/third-party dialogs. . Even though you can change all dialogs in the app to use Radix's Dialog, you can't control how third-party dialogs will render. [...] Here's how we fixed this on Ariakit with a brief explanation: https://github.com/ariakit/ariakit/pull/2339

I'll look into this and report back. As I stated above, removing modality will cause another issue: radix's (or other third-party libs) popover-based components won't work well with anchors in the iframed editor. In our first-party `Popover` component we fixed that by applying some custom middleware to `floating-ui`, and to apply that same "fix" for Radix (or any third-party library) we'd need to:
- either add custom npm patch
- manage to get the fix in Radix (or third party library), wait for the release
- manage to get the fix in floating-ui, wait for a release, then hope that Radix (or third party library) updates their dependencies, and then we'll get it


> I'm also curious to learn how this works [...] otherwise things like roving tabindex may not work as expected

My point regarding the `MenuItem` component was more about reducing breaking changes to a minimum when the legacy `MenuItem` component is rendered inside the new v2 `DropdownMenu` component via slotfill.

The point that you make about internal context not being passed down correctly makes sense though, and would also apply to using the new menu items from the new `DropdownMenu` component.

>  Is migrating all slots to bubblesVirtually still the plan? If so, things might break for third-party components that depend on an internal React context like Radix.

I'm not 100% sure if it is, but I assume that if we got Radix (or third-party libraries) to export their context, we could simply forward them  via `fillProps` like we already do with other contexts?

> I created this example using Ariakit Menu and the Modal component from @wordpress/components that works with SlotFill independently of the bubblesVirtually prop

Is the reason why it works because data is shared via ariakit stores instead of React Context?

diegohaz commented:
> I'll look into this and report back. As I stated above, removing modality will cause another issue: radix's (or other third-party libs) popover-based components won't work well with anchors in the iframed editor. In our first-party `Popover` component we fixed that by applying some custom middleware to `floating-ui`

Where can I learn more about that? Is there a minimum reproduction of that behavior somewhere so I can play with?

> I'm not 100% sure if it is, but I assume that if we got Radix (or third-party libraries) to export their context, we could simply forward them via `fillProps` like we already do with other contexts?

Yes. But that means they will either have to expose their internals or design a public API around that. The first option will put a burden on consumers that will rely on a very unstable code. The latter puts a burden on the library maintainers as designing that kind of API may take a lot of time (it took me years).

> Is the reason why it works because data is shared via ariakit stores instead of React Context?

For the SlotFill thing, yes.

ciampo commented:
> Where can I learn more about that? Is there a minimum reproduction of that behavior somewhere so I can play with?

Last year the `Popover` component [was refactored to using `floating-ui`](https://github.com/WordPress/gutenberg/pull/40740), causing a very high number of regressions that took months to fix (see [this tracking issue](https://github.com/WordPress/gutenberg/issues/42770) for a few examples).

I had also opened a few issues on `floating-ui`'s repository, the main one being [this issue about cross-document support](https://github.com/floating-ui/floating-ui/issues/1869).

On our first party `Popover`, we do a few things to improve the situation when the anchor if in the iframe and the floating is in the iframe's parent document:

- we add [a custom offset middleware](https://github.com/WordPress/gutenberg/blob/b873f437e4f34370f02cf4b99cc90530911b8a9b/packages/components/src/popover/index.tsx#L263-L284) to take into account the offset of the iframe from the document's origin
- We use [a slightly modified version of the `limitShift` middleware](https://github.com/WordPress/gutenberg/blob/b873f437e4f34370f02cf4b99cc90530911b8a9b/packages/components/src/popover/index.tsx#L308), which takes into account the iframe offset that we passed in the custom offset middleware mentioned in the point above
- we [add extra event listeners](https://github.com/WordPress/gutenberg/blob/b873f437e4f34370f02cf4b99cc90530911b8a9b/packages/components/src/popover/index.tsx#L423-L460) for resize and scroll events happening in the iframe's parent document

> Yes. But that means they will either have to expose their internals or design a public API around that. The first option will put a burden on consumers that will rely on a very unstable code. The latter puts a burden on the library maintainers as designing that kind of API may take a lot of time (it took me years).

Couldn't the `Context` be exposed, while leaving the contents of the context as a private implementation detail? Consumers would just need to know that if they need to forward the context, they can do so without worrying about what data is included.

richtabor commented:
Just following up on this; was there any movement since June?

ciampo commented:
> Just following up on this; was there any movement since June?

Hey @richtabor , I have been AFK for the past 2 months and therefore wasn't able to continue work on this.

At the moment, it looks like the experimental Radix UI-based DropdownMenu is not meeting our requirements (see https://github.com/WordPress/gutenberg/pull/51400#issuecomment-1594440489 for more details). I even opened an issue on the Radix UI repo, but 3 months have passed and no one from the project's maintainers has even replied.

In the upcoming weeks, I will try to test Ariakit's DropdownMenu component for this very same test case (the block editor settings dropdown), to see if hopefully it will meet our requirements.

tyxla commented:
@ciampo should we drop this now that we're not planning to continue with Radix as an alternative for the `DropdownMenu` implementation?

ciampo commented:
Yes. I will still refer to this PR when applying the same changes with the ariakit-based dropdown menu, but in the meantime we can close this PR.
